### ISSUE TITLE
Implement new geolocation mechanism

### ISSUE THREAD

**Issue by romac** [reminisceme/app#127]

**Comment by Rabyss** [comment]
The geolocation questions will look like this:
```
{
    userId: String,
    kind: QuestionKind,
    type: QuestionType,
    subject: Subject,
    answer: Location,
}
```
The answer field is a Location:
```
{
    latitude: Number,
    longitude: Number
}
```

To verify an answer, the strategy should be as follows:
Send a geocoding request to Google Maps API:
```
https://maps.googleapis.com/maps/api/geocode/json?latlng=latitude,longitude&key=API_KEY
```
for the correct answer and the user guess. The output of such a request contains an array of results, usually the first one is the most precise one and should be used. Each of the results is an array of `address_component`. An `address_component` is the following JSON:
```
{
    long_name: String,
    short_name: String,
    types: Array[Tag]
}
```
The interesting components contain one of those tags (ordered from least specific to most specific):
1. country
2. administrative_area_level_1
3. administrative_area_level_2
4. administrative_area_level_3
5. administrative_area_level_4
6. administrative_area_level_5
7. locality/ward (indicates a specific type of Japanese locality)
8. sublocality OR  sublocality_level_1 to sublocality_level_5.
9. neighborhood
10. premise
11. subpremise
12. street_number
13. route
(for more information on the tags: https://developers.google.com/maps/documentation/geocoding/intro?csw=1#Types)
Then the correctness is evaluated as follows:
1. Take the most specific result for both the user guess and the correct answer (should be the first one)
2. Check if the user guess is as specific as the correct answer, i.e. their most specific component is on the same level (levels 9 to 13 may be too specific to be fair so one might ignore them)
2. Check if the names of the components 1 to most specific are the same

(Sample Google Map API output:
```
https://maps.googleapis.com/maps/api/geocode/json?latlng=40.714224,-73.961452
```
Answer:
```
{
results: [
{
address_components: [
{
long_name: "277",
short_name: "277",
types: [
"street_number"
]
},
{
long_name: "Bedford Avenue",
short_name: "Bedford Ave",
types: [
"route"
]
},
{
long_name: "Williamsburg",
short_name: "Williamsburg",
types: [
"neighborhood",
"political"
]
},
{
long_name: "Brooklyn",
short_name: "Brooklyn",
types: [
"sublocality_level_1",
"sublocality",
"political"
]
},
{
long_name: "Kings County",
short_name: "Kings County",
types: [
"administrative_area_level_2",
"political"
]
},
{
long_name: "New York",
short_name: "NY",
types: [
"administrative_area_level_1",
"political"
]
},
{
long_name: "États-Unis",
short_name: "US",
types: [
"country",
"political"
]
},
{
long_name: "11211",
short_name: "11211",
types: [
"postal_code"
]
}
],
formatted_address: "277 Bedford Ave, Brooklyn, NY 11211, États-Unis",
geometry: {
location: {
lat: 40.714232,
lng: -73.9612889
},
location_type: "ROOFTOP",
viewport: {
northeast: {
lat: 40.7155809802915,
lng: -73.9599399197085
},
southwest: {
lat: 40.7128830197085,
lng: -73.96263788029151
}
}
},
place_id: "ChIJd8BlQ2BZwokRAFUEcm_qrcA",
types: [
"street_address"
]
},
{
address_components: [
{
long_name: "Grand St/Bedford Av",
short_name: "Grand St/Bedford Av",
types: [
"point_of_interest",
"establishment"
]
},
{
long_name: "Williamsburg",
short_name: "Williamsburg",
types: [
"neighborhood",
"political"
]
},
{
long_name: "Brooklyn",
short_name: "Brooklyn",
types: [
"sublocality_level_1",
"sublocality",
"political"
]
},
{
long_name: "Kings County",
short_name: "Kings County",
types: [
"administrative_area_level_2",
"political"
]
},
{
long_name: "New York",
short_name: "NY",
types: [
"administrative_area_level_1",
"political"
]
},
{
long_name: "États-Unis",
short_name: "US",
types: [
"country",
"political"
]
},
{
long_name: "11211",
short_name: "11211",
types: [
"postal_code"
]
}
],
formatted_address: "Grand St/Bedford Av, Brooklyn, NY 11211, États-Unis",
geometry: {
location: {
lat: 40.714321,
lng: -73.961151
},
location_type: "APPROXIMATE",
viewport: {
northeast: {
lat: 40.71566998029149,
lng: -73.95980201970849
},
southwest: {
lat: 40.7129720197085,
lng: -73.96249998029151
}
}
},
place_id: "ChIJi27VXGBZwokRM8ErPyB91yk",
types: [
"bus_station",
"transit_station",
"point_of_interest",
"establishment"
]
},
{
address_components: [
{
long_name: "Williamsburg",
short_name: "Williamsburg",
types: [
"neighborhood",
"political"
]
},
{
long_name: "Brooklyn",
short_name: "Brooklyn",
types: [
"sublocality_level_1",
"sublocality",
"political"
]
},
{
long_name: "New York",
short_name: "NY",
types: [
"locality",
"political"
]
},
{
long_name: "Kings County",
short_name: "Kings County",
types: [
"administrative_area_level_2",
"political"
]
},
{
long_name: "État de New York",
short_name: "NY",
types: [
"administrative_area_level_1",
"political"
]
},
{
long_name: "États-Unis",
short_name: "US",
types: [
"country",
"political"
]
}
],
formatted_address: "Williamsburg, Brooklyn, État de New York, États-Unis",
geometry: {
bounds: {
northeast: {
lat: 40.7251773,
lng: -73.936498
},
southwest: {
lat: 40.6979329,
lng: -73.96984499999999
}
},
location: {
lat: 40.7081156,
lng: -73.9570696
},
location_type: "APPROXIMATE",
viewport: {
northeast: {
lat: 40.7251773,
lng: -73.936498
},
southwest: {
lat: 40.6979329,
lng: -73.96984499999999
}
}
},
place_id: "ChIJQSrBBv1bwokRbNfFHCnyeYI",
types: [
"neighborhood",
"political"
]
},
{
address_components: [
{
long_name: "Brooklyn",
short_name: "Brooklyn",
types: [
"sublocality_level_1",
"sublocality",
"political"
]
},
{
long_name: "New York",
short_name: "NY",
types: [
"locality",
"political"
]
},
{
long_name: "Kings County",
short_name: "Kings County",
types: [
"administrative_area_level_2",
"political"
]
},
{
long_name: "État de New York",
short_name: "NY",
types: [
"administrative_area_level_1",
"political"
]
},
{
long_name: "États-Unis",
short_name: "US",
types: [
"country",
"political"
]
}
],
formatted_address: "Brooklyn, État de New York, États-Unis",
geometry: {
bounds: {
northeast: {
lat: 40.739446,
lng: -73.8333651
},
southwest: {
lat: 40.551042,
lng: -74.05663
}
},
location: {
lat: 40.6781784,
lng: -73.9441579
},
location_type: "APPROXIMATE",
viewport: {
northeast: {
lat: 40.739446,
lng: -73.8333651
},
southwest: {
lat: 40.551042,
lng: -74.05663
}
}
},
place_id: "ChIJCSF8lBZEwokRhngABHRcdoI",
types: [
"sublocality_level_1",
"sublocality",
"political"
]
},
{
address_components: [
{
long_name: "New York",
short_name: "NY",
types: [
"locality",
"political"
]
},
{
long_name: "État de New York",
short_name: "NY",
types: [
"administrative_area_level_1",
"political"
]
},
{
long_name: "États-Unis",
short_name: "US",
types: [
"country",
"political"
]
}
],
formatted_address: "New York, État de New York, États-Unis",
geometry: {
bounds: {
northeast: {
lat: 40.91525559999999,
lng: -73.70027209999999
},
southwest: {
lat: 40.4960439,
lng: -74.2557349
}
},
location: {
lat: 40.7127837,
lng: -74.0059413
},
location_type: "APPROXIMATE",
viewport: {
northeast: {
lat: 40.91525559999999,
lng: -73.70027209999999
},
southwest: {
lat: 40.4960439,
lng: -74.2557349
}
}
},
place_id: "ChIJOwg_06VPwokRYv534QaPC8g",
types: [
"locality",
"political"
]
},
{
address_components: [
{
long_name: "11211",
short_name: "11211",
types: [
"postal_code"
]
},
{
long_name: "Brooklyn",
short_name: "Brooklyn",
types: [
"sublocality_level_1",
"sublocality",
"political"
]
},
{
long_name: "New York",
short_name: "NY",
types: [
"locality",
"political"
]
},
{
long_name: "État de New York",
short_name: "NY",
types: [
"administrative_area_level_1",
"political"
]
},
{
long_name: "États-Unis",
short_name: "US",
types: [
"country",
"political"
]
}
],
formatted_address: "Brooklyn, État de New York 11211, États-Unis",
geometry: {
bounds: {
northeast: {
lat: 40.7280089,
lng: -73.9207299
},
southwest: {
lat: 40.7008331,
lng: -73.9644697
}
},
location: {
lat: 40.7093358,
lng: -73.9565551
},
location_type: "APPROXIMATE",
viewport: {
northeast: {
lat: 40.7280089,
lng: -73.9207299
},
southwest: {
lat: 40.7008331,
lng: -73.9644697
}
}
},
place_id: "ChIJvbEjlVdZwokR4KapM3WCFRw",
types: [
"postal_code"
]
},
{
address_components: [
{
long_name: "Kings County",
short_name: "Kings County",
types: [
"administrative_area_level_2",
"political"
]
},
{
long_name: "New York",
short_name: "NY",
types: [
"administrative_area_level_1",
"political"
]
},
{
long_name: "États-Unis",
short_name: "US",
types: [
"country",
"political"
]
}
],
formatted_address: "Kings County, NY, États-Unis",
geometry: {
bounds: {
northeast: {
lat: 40.7391407,
lng: -73.8333651
},
southwest: {
lat: 40.5702197,
lng: -74.0419592
}
},
location: {
lat: 40.6528762,
lng: -73.95949399999999
},
location_type: "APPROXIMATE",
viewport: {
northeast: {
lat: 40.7391407,
lng: -73.8333651
},
southwest: {
lat: 40.5702197,
lng: -74.0419592
}
}
},
place_id: "ChIJOwE7_GTtwokRs75rhW4_I6M",
types: [
"administrative_area_level_2",
"political"
]
},
{
address_components: [
{
long_name: "État de New York",
short_name: "NY",
types: [
"administrative_area_level_1",
"political"
]
},
{
long_name: "États-Unis",
short_name: "US",
types: [
"country",
"political"
]
}
],
formatted_address: "État de New York, États-Unis",
geometry: {
bounds: {
northeast: {
lat: 45.015865,
lng: -71.8562293
},
southwest: {
lat: 40.496091,
lng: -79.76214379999999
}
},
location: {
lat: 43.2994285,
lng: -74.21793260000001
},
location_type: "APPROXIMATE",
viewport: {
northeast: {
lat: 45.015859,
lng: -71.8562293
},
southwest: {
lat: 40.496091,
lng: -79.76214379999999
}
}
},
place_id: "ChIJqaUj8fBLzEwRZ5UY3sHGz90",
types: [
"administrative_area_level_1",
"political"
]
},
{
address_components: [
{
long_name: "États-Unis",
short_name: "US",
types: [
"country",
"political"
]
}
],
formatted_address: "États-Unis",
geometry: {
bounds: {
northeast: {
lat: 71.3867745,
lng: -66.9502861
},
southwest: {
lat: 18.9106768,
lng: 172.4458955
}
},
location: {
lat: 37.09024,
lng: -95.712891
},
location_type: "APPROXIMATE",
viewport: {
northeast: {
lat: 49.38,
lng: -66.94
},
southwest: {
lat: 25.82,
lng: -124.39
}
}
},
place_id: "ChIJCzYy5IS16lQRQrfeQ5K5Oxw",
types: [
"country",
"political"
]
}
],
status: "OK"
}
```

**Comment by Aranir** [comment]
implemented on the app with commit: @aac429095b535f2a81751554fc58bda8c496e01f
